src/rgw/rgw_gc.cc

```diff
@@ -90,15 +101,75 @@ int RGWGC::defer_chain(const string& tag, cls_rgw_obj_chain& chain, bool sync)
   info.chain = chain;
   info.tag = tag;
 
-  ObjectWriteOperation op;
-  cls_rgw_gc_queue_defer_entry(op, cct->_conf->rgw_gc_obj_min_wait, info);
-
   int i = tag_index(tag);
 
-  if (sync)
-    return store->gc_operate(obj_names[i], &op);
+  ObjectWriteOperation op;
```

Comment:
@cbodley @mattbenjamin : If we want to process omap entries first, mark the transition as complete and then process queue entries, then we have to make sure that all entries in omap expire before the queue entries. For this we can't have a defer op go to the omap, it may set expiration time of an entry to a time later than entries in the queue. To simplify things I have added this logic in defer_gc: Remove the tag from omap, if it exists there, and then add an entry to the queue. (can we have the remove and enqueue steps as part of a single transaction?)

Reply:
> can we have the remove and enqueue steps as part of a single transaction?

yes, so long as they refer to the same rados object

Reply:
Ok

Reply:
that's allowed, yeah. what if we do the enqueue first, followed by the omap remove? that way if the enqueue fails with EPERM, the omap remove part of the transaction doesn't get executed. and defer_chain() can handle the EPERM case by falling back to the original omap defer op. if transitioned_objects_cache is set, you just skip the omap remove part
```c++
ObjectWriteOperation op;

// try cls_rgw_gc defer
cls_rgw_gc_queue_defer_entry(op, cct->_conf->rgw_gc_obj_min_wait, info);
if (!transitioned_objects_cache[i]) {
  // remove from cls_rgw's gc so it doesn't run until we transition
  cls_rgw_gc_remove(op, tags);
}
int ret = store->gc_operate(obj_names[i], &op);

if (ret == -EPERM) { // cls_rgw_gc not available, fall back to cls_rgw defer
  ObjectWriteOperation fallback_op;
  cls_rgw_gc_defer_entry(fallback_op, cct->_conf->rgw_gc_obj_min_wait, tag);
  ret = store->gc_operate(obj_names[i], &fallback_op);
}
```
the async case is tricky though, since we don't wait for a reply from the osd. that means you can't rely on the return value of gc_aio_operate() to see the EPERM error. since gc_aio_operate() calls the librados aio_operate(), we'd have to add an AioCompletion that handles the EPERM fallback